Add fixed gas price and valid until (#22)
OpenZeppelin/defender-docs: docs/modules/ROOT/pages/relay.adoc

[[speed]]
=== Speed

Instead of the usual `gasPrice`, the Relayer accepts a <<relay-api-reference.adoc#send-transaction,speed>> parameter, which can be `safeLow`, `average`, `fast`, or `fastest`. These values are mapped to actual gas prices when the transaction is sent or resubmitted and vary depending on the state of the network.

If have to use a **fixed gas price** for a transaction to send, it can be set using a <<relay-api-reference.adoc#send-transaction,gasPrice>> paramter. Defender Relay will never change a fixed gas price and transaction either will be mined with a provided `gasPrice` or replaced by a NOOP tx if it is not mined before <<valid-until, validUntil>> time.

Keep in mind that you have to provide either `speed` or `gasPrice` but not both in a send transaction request.

NOTE: Today, mainnet gas prices for each speed are obtained from https://ethgasstation.info/[ethgasstation], falling back to https://etherchain.org/tools/gasPriceOracle[etherchain] if the former is not available. We are experimenting with other gas price oracles in the meantime. Testnet gas prices are obtained from a call to `eth_gasPrice` to the network.

[[valid-until]]
=== Valid Until

> Every transaction in Defender Relay is valid for submission to the Ethereum network until <<relay-api-reference.adoc#send-transaction,validUntil>> time. After `validUntil` time the transaction is replaced by a NOOP transaction in order to prevent Relayer from being stuck on the transaction's nonce. A NOOP transaction does nothing except advancing relayer's nonce.

Reviewer: Relayer => the relayer

Reviewer: from being stuck => from getting stuck

(and I think it should be `getting stuck at` instead of `on`, but I'm not sure!)

Reviewer: does nothing except advancing relayer's nonce => does nothing except advancing the relayer's nonce

> A default `validUntil` value of a transaction equals to time of creation plus **8 hours**; meaning the transaction have 8 hours to be mined. You can set this parameter manually to be more aggresive on how fast your transaction can be mined. This can be useful combined with a fixed gas price to archive extremely fast minnig time and beating other txs on `gasPrice`.

Reviewer: This feels a bit redundant. After your explanation above I think you could just say: "`validUntils` defaults to 8 hours after transaction creation."

Then:

"Note that you can combine `validUntil` with a fixed gas price to achieve extremely fast mining times and beating other transactions on `gasPrice`."

[[speed]]
=== Speed

Instead of the usual `gasPrice`, the Relayer accepts a <<relay-api-reference.adoc#send-transaction,speed>> parameter, which can be `safeLow`, `average`, `fast`, or `fastest`. These values are mapped to actual gas prices when the transaction is sent or resubmitted and vary depending on the state of the network.

NOTE: Today, mainnet gas prices for each speed are obtained from https://ethgasstation.info/[ethgasstation], falling back to https://etherchain.org/tools/gasPriceOracle[etherchain] if the former is not available. We are experimenting with other gas price oracles in the meantime. Testnet gas prices are obtained from a call to `eth_gasPrice` to the network.

[[gas-price]]
=== Gas Price

> Alternatively, you may specify a **fixed gas price** for a transaction, by setting the <<relay-api-reference.adoc#send-transaction,gasPrice>> parameter. Defender Relay will never change a fixed gas price and transaction either will be mined with a provided `gasPrice` or replaced by a NOOP tx if it is not mined before <<valid-until, validUntil>> time.

Reviewer: Something sounds odd about the last sentence. How about...

"Transactions with fixed gas price are either mined with the specified `gasPrice` or replaced with a NOOP transaction if they couldn't be mined before <<valid-until, validUntil>> time."

> Alternatively, you may specify a **fixed gas price** for a transaction, by setting the <<relay-api-reference.adoc#send-transaction,gasPrice>> parameter. Defender Relay will never change a fixed gas price and transaction either will be mined with a provided `gasPrice` or replaced by a NOOP tx if it is not mined before <<valid-until, validUntil>> time.

Author: ```suggestion
Alternatively, you may specify a **fixed gas price** for a transaction, by setting the <<relay-api-reference.adoc#send-transaction,gasPrice>> parameter.
Transactions with a fixed gas price are either mined with the specified `gasPrice` or replaced with a NOOP transaction if they couldn't be mined before <<valid-until, validUntil>> time.
```

Keep in mind that you have to provide either `speed` or `gasPrice` but not both in a send transaction request.

[[valid-until]]
=== Valid Until

Every transaction in Defender Relay is valid for submission to the Ethereum network until <<relay-api-reference.adoc#send-transaction,validUntil>> time. After `validUntil` time the transaction is replaced by a NOOP transaction in order to prevent relayer from getting stuck at the transaction's nonce. A NOOP transaction does nothing except advancing the relayer's nonce.

> `validUntils` defaults to 8 hours after a transaction creation. Note that you can combine validUntil with a fixed gas price to achieve extremely fast mining times and beating other transactions on gasPrice.

Reviewer: `a transaction creation` => `the transaction creation`  😬 

I think you missed backticks around the last `gasPrice` mention

Author: ```suggestion
`validUntil` defaults to 8 hours after the transaction creation. Note that you can combine validUntil with a fixed gas price to achieve extremely fast mining times and beating other transactions on `gasPrice`.
```

`validUntil` defaults to 8 hours after the transaction creation. Note that you can combine validUntil with a fixed gas price to achieve extremely fast mining times and beating other transactions on `gasPrice`.

> In order to use `validUntil` with `ethers.js` provide an option `validForSeconds` for `DefenderRelaySigner` there `validForSecods` stands for how many seconds the transaction would be valid after its creation.

Reviewer: If you're using `ethers.js`, you may set a `validForSeconds` option instead of `validUntil`. In the example below, we configure a `DefenderRelaySigner` to issue a transaction which will be valid for 120 seconds after its creation.

> In order to use `validUntil` with `ethers.js` provide an option `validForSeconds` for `DefenderRelaySigner` there `validForSecods` stands for how many seconds the transaction would be valid after its creation.

Author: ```suggestion
If you're using `ethers.js`, you may set a `validForSeconds` option instead of `validUntil`. In the example below, we configure a `DefenderRelaySigner` to issue a transaction which will be valid for 120 seconds after its creation.
```

[source,jsx]
----
const signer = new DefenderRelaySigner(credentials, provider, { validForSeconds: 120 });
----

> NOTE: `validUntil` is a UTC timestamp. Make sure to use a UTC timezone not a local one.

Reviewer: NOTE: `validUntil` is a UTC timestamp. Make sure to use a UTC timezone and not a local one.

> NOTE: `validUntil` is a UTC timestamp. Make sure to use a UTC timezone not a local one.

Author: ```suggestion
NOTE: `validUntil` is a UTC timestamp. Make sure to use a UTC timezone and not a local one.
```
